lib/at_cmd/at_cmd.c

```diff
@@ -162,133 +230,126 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+			memcpy(current_cmd.resp, buf, payload_len);
 		}
 
-		if (payload_len == 0) {
-			goto next;
+		/* Call the relevant callback, if any */
+		if (ret.state == AT_CMD_NOTIFICATION &&
+		    notification_handler != NULL) {
+			notification_handler(buf);
+		} else if (current_cmd.callback != NULL) {
+			current_cmd.callback(buf);
 		}
 
-		if (ret.state == AT_CMD_NOTIFICATION) {
-			item->callback = notification_handler;
-		} else {
-			item->callback = current_cmd_handler;
-		}
 next:
-		/* If no callback was set, free the item.
-		 * Otherwise, work queue callback will free it.
-		 */
-		if (item->callback == NULL) {
-			k_mem_slab_free(&rsp_work_items, (void **)&item);
-		} else {
-			k_work_init(&item->work, callback_worker);
-			k_work_submit(&item->work);
+		/* Dispatch response for sync call */
+		if (current_cmd.cmd != NULL &&
+		    current_cmd.flags & AT_CMD_SYNC &&
+		    ret.state != AT_CMD_NOTIFICATION) {
+			LOG_DBG("Enqueueing response for sync call");
+			k_msgq_put(&response_sync, &ret, K_FOREVER);
 		}
 
-		/* Notify back only if command was sent. */
-		if ((k_sem_count_get(&cmd_pending) == 0) &&
-		    (ret.state != AT_CMD_NOTIFICATION)) {
-			current_cmd_handler = NULL;
-
-			k_msgq_put(&return_code_msq, &ret, K_FOREVER);
+		/* We have now handled a command if it was not a notification */
+		if (ret.state != AT_CMD_NOTIFICATION) {
+			complete_cmd();
 		}
 	}
 }
 
-static inline int at_write(const char *const cmd, enum at_cmd_state *state)
+int at_cmd_write_with_callback(const char *const cmd,
+			       at_cmd_handler_t  handler)
 {
-	int bytes_sent;
-	int bytes_to_send = strlen(cmd);
-	struct return_state_object ret;
+	struct cmd_item command;
+	int ret;
 
-	LOG_DBG("Sending command %s", log_strdup(cmd));
-
-	bytes_sent = send(common_socket_fd, cmd, bytes_to_send, 0);
-
-	if (bytes_sent == -1) {
-		LOG_ERR("Failed to send AT command (err:%d)", errno);
-		ret.code  = -errno;
-		ret.state = AT_CMD_ERROR;
-	} else {
-		LOG_DBG("Awaiting response for %s", log_strdup(cmd));
-		k_msgq_get(&return_code_msq, &ret, K_FOREVER);
-		LOG_DBG("Bytes sent: %d", bytes_sent);
-
-		if (bytes_sent != bytes_to_send) {
-			LOG_ERR("Bytes sent (%d) was not the "
-				"same as expected (%d)",
-				bytes_sent, bytes_to_send);
-		}
+	if (cmd == NULL) {
+		return -EINVAL;
 	}
 
-	if (state) {
-		*state = ret.state;
+	command.cmd = k_malloc(strlen(cmd) + 1);
+	if (command.cmd == NULL) {
+		return -ENOMEM;
 	}
+	strcpy(command.cmd, cmd);
 
-	return ret.code;
-}
-
-int at_cmd_write_with_callback(const char *const cmd,
-			       at_cmd_handler_t  handler,
-			       enum at_cmd_state *state)
-{
-	k_sem_take(&cmd_pending, K_FOREVER);
-
-	current_cmd_handler = handler;
+	command.resp = NULL;
+	command.callback = handler;
+	command.flags = AT_CMD_BUF_CMD;
 
-	int return_code = at_write(cmd, state);
-
-	k_sem_give(&cmd_pending);
+	ret = k_msgq_put(&commands, &command, K_FOREVER);
+	if (ret) {
+		return ret;
+	}
 
-	return return_code;
+	load_cmd_and_write();
```

Comment:
Doesn't this mean that you will always process the queue with exactly one command?

Reply:
I understand the confusion; `load_cmd_and_write` does not necessarily load a cmd and write it. As the description hints, it only does that "if appropriate". That is, only if no command is currently being processed. The logic can be seen in the if statement on line 168.

Reply:
In effect, that function is a synchronization mechanism that avoids having two separate threads for listening to the queue and the socket.